### PR TITLE
Fixes NullRef exception in IIS in-proc

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/dotnet/aspnetcore/issues/17516

#### Description

Customers are hitting a null ref exception occasionally on requests closing. The fix is to fix a couple race conditions in the IIS in-proc code.

#### Customer Impact

Customers reported the issue in https://github.com/dotnet/aspnetcore/issues/17516 and there have been many comments on it about hitting the same issue (+11 on one comment).

Exceptions show up in debugger. And this has increased in frequency with VS 16.5

No workarounds besides ignoring the exception.

#### Regression?

No, this has been there since in-proc was released.

#### Risk

Low, mostly just extra null checks.

### CODE REVIEW

File: src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp [hunk 1]
```diff
@@ -139,14 +139,16 @@ IN_PROCESS_HANDLER::SetManagedHttpContext(
     PVOID pManagedHttpContext
 )
 {
+    bool disconnectFired = false;
```
Comment (halter73) [inline src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp:142]: nit:

```suggestion
    bool disconnectFired = false;

```

File: src/Servers/IIS/IIS/src/Core/IISHttpContext.cs [hunk 2]
```diff
@@ -119,71 +119,77 @@ internal abstract partial class IISHttpContext : NativeRequestContext, IThreadPo
...
+                    new PipeOptions(
+                        _memoryPool,
+                        readerScheduler: PipeScheduler.ThreadPool,
+                        pauseWriterThreshold: PauseWriterThreshold,
+                        resumeWriterThreshold: ResumeWriterTheshold,
+                        minimumSegmentSize: MinAllocBufferSize));
+                _bodyOutput = new OutputProducer(pipe);
 
-            NativeMethods.HttpSetManagedContext(_pInProcessHandler, (IntPtr)_thisHandle);
+                NativeMethods.HttpSetManagedContext(_pInProcessHandler, (IntPtr)_thisHandle);
```
Comment (halter73) [inline src/Servers/IIS/IIS/src/Core/IISHttpContext.cs:191]: Nit: Moving HttpSetManagedContext after the lock avoids possible (rare) contention.

**Review by halter73 — APPROVED** [review]

**Review by jkotalik — APPROVED** [review]

**Comment by Pilchie** [comment]
@BrennanConroy want me to squash and merge.

**Comment by BrennanConroy** [comment]
> @BrennanConroy want me to squash and merge.

Yes please 😄

**Pull request merged**